File: ios/Sources/AmthalPortal/PortalConfig.swift
import Foundation
import UIKit

/// Portal locale, forwarded to the web app via `configure`.
/// `ar` also flips document direction to RTL on the web side.
public enum PortalLocale: String, Codable, CaseIterable, Sendable {
    case en
    case ar
}

/// Portal theme, forwarded to the web app via `configure`.
public enum PortalTheme: String, Codable, CaseIterable, Sendable {
    case light
    case dark
    case system
}

/// Errors thrown by `PortalConfig.init` when the base URL is unacceptable.
public enum PortalConfigError: Error, Equatable, LocalizedError {
    /// The base URL has no usable scheme/host.
    case invalidBaseURL(URL)
    /// `http:` base URLs are rejected — HTTPS only (exception: localhost for development).
    case insecureBaseURL(URL)

    public var errorDescription: String? {
        switch self {
        case .invalidBaseURL(let url):
            return "Invalid portal base URL: \(url.absoluteString)"
        case .insecureBaseURL(let url):
            return "Portal base URL must use https (http is only allowed for localhost, "
                + "or any host via the DEV-only allowInsecureHttp flag): \(url.absoluteString)"
        }
    }
}

/// Immutable-origin configuration for a portal session.
public struct PortalConfig: Equatable, Sendable {
    /// Portal base URL, e.g. `https://portal.example.com` — **including any
    /// deploy path prefix** when the portal is served under a baseHref, e.g.
    /// `https://host.example.com/portal`. HTTPS required (`http` allowed for
    /// localhost only, for development).
    public let baseURL: URL

    /// Path component of `baseURL` without trailing slash (`""` when the
    /// portal is deployed at the origin root). Mirror of `basePathOf()` in
    /// protocol.ts; all embed-subtree checks are made against
    /// `basePath + "/embed"`.
    public let basePath: String

    /// UI locale passed to the portal (`configure.locale`).
    public var locale: PortalLocale

    /// UI theme passed to the portal (`configure.theme`).
    public var theme: PortalTheme

    /// 1.0 = default; the portal maps it to a root font-size adjustment.
    public var fontScale: Double

    /// Your brand colour, used for the SDK's OWN interstitials (the loading indicator and
    /// the retry affordance) — not for the portal's UI, which is themed server-side.
    ///
    /// It cannot arrive over the bridge: the interstitial is on screen before the WebView
    /// can speak, so nothing the portal sends could colour the first frame. Set it here and
    /// the wait is your colour from the very first pixel.
    ///
    /// `nil` (the default) draws with the platform's own label colour rather than a colour
    /// we invented — deliberately plain, never another tenant's brand.
    public var brandColor: UIColor?

    /// When set, the manifest's `portalVersion` must be >= this or the SDK
    /// surfaces `versionIncompatible`.
    public var minPortalVersion: String?

    /// Hosts the WebView may navigate to (main frame) and that bridge traffic
    /// is accepted from. Defaults to the `baseURL` host.
    public var allowedOrigins: Set<String>

    /// Enables SDK diagnostics logging (DEBUG builds only; never logs
    /// credentials or header values).
    public var enableLogging: Bool

    /// DEV ONLY: permit a plain-`http` base URL for non-localhost hosts —
    /// needed when the portal dev server is bound to a tenant hostname (the
    /// backend resolves the tenant from the request host/referer). The host
    /// app also needs an ATS exception for the domain.
    ///
    /// This is what was *honoured*, not what the caller asked for: it is
    /// forced to `false` outside DEBUG builds (see `honourInsecureHTTP`), so
    /// `isOriginAllowed` and the init check below are gated by construction.
    public let allowInsecureHTTP: Bool

    public init(
        baseURL: URL,
        locale: PortalLocale = .en,
        theme: PortalTheme = .system,
        fontScale: Double = 1.0,
        brandColor: UIColor? = nil,
        minPortalVersion: String? = nil,
        allowedOrigins: Set<String>? = nil,
        enableLogging: Bool = false,
        allowInsecureHTTP: Bool = false
    ) throws {
        guard let scheme = baseURL.scheme?.lowercased(),
              let host = baseURL.host, !host.isEmpty
        else {
            throw PortalConfigError.invalidBaseURL(baseURL)
        }
        let insecureHTTPHonoured = Self.honourInsecureHTTP(allowInsecureHTTP)
        switch scheme {
        case "https":
            break
        case "http" where Self.isLocalhost(host) || insecureHTTPHonoured:
            break
        default:
            throw PortalConfigError.insecureBaseURL(baseURL)
        }
        self.baseURL = baseURL
        // basePathOf(): path without trailing slash, '' for root.
        var basePath = baseURL.path
        while basePath.hasSuffix("/") { basePath.removeLast() }
        self.basePath = basePath
        self.locale = locale
        self.theme = theme
        self.fontScale = fontScale
        self.brandColor = brandColor
        self.minPortalVersion = minPortalVersion
        self.allowedOrigins = allowedOrigins ?? [host]
        self.enableLogging = enableLogging
        self.allowInsecureHTTP = insecureHTTPHonoured
    }

    // MARK: Internal helpers

    /// The DEV-only cleartext escape hatch counts only in a DEBUG build: a
    /// shipped app must not be talked into http by one boolean in the
    /// embedding code. Outside DEBUG the flag is dropped, and a base URL that
    /// needed it then fails `init` with `.insecureBaseURL` — a thrown, typed
    /// error rather than a silent downgrade.
    private static func honourInsecureHTTP(_ requested: Bool) -> Bool {
        #if DEBUG
        return requested
        #else
        return false
        #endif
    }

    static func isLocalhost(_ host: String) -> Bool {
        let h = host.lowercased()
        return h == "localhost" || h == "127.0.0.1" || h == "::1" || h == "[::1]"
            || h.hasSuffix(".localhost")
    }

    /// True when `url`'s origin (scheme + host) is acceptable for bridge
    /// traffic and main-frame navigation.
    func isOriginAllowed(_ url: URL?) -> Bool {
        guard let url,
              let host = url.host, !host.isEmpty,
              let scheme = url.scheme?.lowercased()
        else { return false }
        guard isHostAllowed(host) else { return false }
        if scheme == "https" { return true }
        if scheme == "http", Self.isLocalhost(host) || allowInsecureHTTP { return true }
        return false
    }

    func isHostAllowed(_ host: String) -> Bool {
        allowedOrigins.contains { $0.caseInsensitiveCompare(host) == .orderedSame }
    }

    /// True when `path` (a URL path component) is within the embed subtree of
    /// a portal deployed at `basePath`. Mirror of `isEmbedPath()` in
    /// protocol.ts — the check used for main-frame navigation confinement.
    /// Never test against a literal "/embed" prefix.
    func isEmbedPath(_ path: String) -> Bool {
        let prefix = basePath + BridgeProtocol.embedPathPrefix
        return path == prefix || path.hasPrefix(prefix + "/")
    }

    /// Builds `{baseURL}{path}`, preserving any deploy path prefix in the
    /// base URL. Mirror of `resolveEmbedUrl()`/`manifestUrl()` in protocol.ts
    /// (trailing-slash-stripped base + path). The URL *is* the request —
    /// never put tokens or other credentials in it.
    func portalURL(forPath path: String) -> URL? {
        guard let components = URLComponents(url: baseURL, resolvingAgainstBaseURL: true),
              let scheme = components.scheme,
              let host = components.host
        else { return nil }
        let portPart = components.port.map { ":\($0)" } ?? ""
        // Percent-encoded base path keeps the rebuilt URL string valid;
        // any query/fragment on the base URL is deliberately dropped.
        var encodedBasePath = components.percentEncodedPath
        while encodedBasePath.hasSuffix("/") { encodedBasePath.removeLast() }
        return URL(string: "\(scheme)://\(host)\(portPart)\(encodedBasePath)\(path)")
    }
}

File: android/portal-sdk/src/main/java/com/amthalgroup/portal/PortalConfig.kt
package com.amthalgroup.portal

import android.content.Context
import android.content.pm.ApplicationInfo
import android.net.Uri
import androidx.annotation.ColorInt
import android.util.Log
import java.security.cert.X509Certificate

/** UI locale passed to the portal (bridge `configure.locale`). */
enum class PortalLocale(val wire: String) {
    EN("en"),
    AR("ar")
}

/** Theme passed to the portal (bridge `configure.theme`). */
enum class PortalTheme(val wire: String) {
    LIGHT("light"),
    DARK("dark"),
    SYSTEM("system")
}

/**
 * Immutable configuration for an embedded portal session.
 *
 * @param baseUrl full portal base URL, INCLUDING any deploy path prefix when
 *   the portal is served under a baseHref — e.g. `https://portal.example.com`
 *   or `https://host.example.com/portal`. A trailing slash is tolerated.
 *   Must be `https`; `http` is allowed only for local development hosts
 *   (`localhost`, `127.0.0.1`, `10.0.2.2`, `::1`). No query or fragment.
 * @param locale initial portal locale.
 * @param theme initial portal theme.
 * @param fontScale initial font scale (1.0 = default).
 * @param brandColor ARGB colour for the SDK's OWN interstitials — the loading indicator
 *   and the retry affordance — not for the portal's UI, which is themed server-side from
 *   the branch. It cannot arrive over the bridge: the interstitial is on screen before the
 *   WebView can speak, so nothing the portal sends could colour the first frame. `null`
 *   (the default) leaves the widget to tint itself from the host app's own theme, rather
 *   than substituting a colour of ours — never another tenant's brand.
 * @param minPortalVersion optional floor for the portal version reported by
 *   `/embed/manifest.json`; older portals fail with `VERSION_INCOMPATIBLE`.
 * @param allowedOrigins additional allowed hosts (or full origins) for the
 *   WebView / bridge origin check. When empty, only the [baseUrl] host is
 *   allowed.
 * @param certificatePinner optional hook invoked with the server's leaf
 *   certificate for SDK-initiated HTTPS connections (manifest fetch and
 *   downloads); return `false` to abort the connection. Note: the portal
 *   page load itself goes through the system WebView network stack, which
 *   does not expose a pinning hook — pin at the network edge if you need
 *   full coverage.
 * @param enableLogging when `true`, bridge `log` messages and WebView console
 *   output are forwarded to Logcat. Keep `false` in release builds.
 * @param allowInsecureHttp DEV ONLY: permit a plain-`http` [baseUrl] for
 *   non-loopback hosts — needed when the portal dev server is bound to a
 *   tenant hostname (the backend resolves the tenant from the request
 *   host/referer). Honoured only in a debuggable app — see
 *   [isInsecureHttpHonoured]; the app manifest must also allow cleartext
 *   traffic to the dev host.
 */
data class PortalConfig(
    val baseUrl: String,
    val locale: PortalLocale = PortalLocale.EN,
    val theme: PortalTheme = PortalTheme.SYSTEM,
    val fontScale: Float = 1f,
    @ColorInt val brandColor: Int? = null,
    val minPortalVersion: String? = null,
    val allowedOrigins: Set<String> = emptySet(),
    val certificatePinner: ((X509Certificate) -> Boolean)? = null,
    val enableLogging: Boolean = false,
    val allowInsecureHttp: Boolean = false
) {

    /** Precomputed portal origin (scheme://host[:port] — no path, no trailing slash). */
    val origin: String

    /**
     * Path component of [baseUrl] without trailing slash; empty (`""`) for a
     * root deploy, `/portal` for `https://host/portal/`. Mirrors
     * `basePathOf()` in protocol.ts. Every embed-path check tests against
     * `basePath + "/embed"` — never a literal `/embed`.
     */
    val basePath: String

    /**
     * [baseUrl] with trailing slashes stripped — the string all portal URLs
     * (manifest fetch, initial load) are concatenated onto. Mirrors the
     * normalization in protocol.ts `resolveEmbedUrl()` / `manifestUrl()`.
     */
    val normalizedBaseUrl: String

    /** Lower-cased hosts allowed for main-frame navigation and bridge traffic. */
    val allowedHosts: Set<String>

    /** Lower-cased scheme of [baseUrl] — read by [isBaseUrlTransportAllowed]. */
    private val baseScheme: String

    /** Lower-cased host of [baseUrl] — read by [isBaseUrlTransportAllowed]. */
    private val baseHost: String

    init {
        require(fontScale > 0f) { "fontScale must be > 0" }
        val uri = Uri.parse(baseUrl.trim())
        val scheme = uri.scheme?.lowercase()
            ?: throw IllegalArgumentException("baseUrl must be an absolute URL: $baseUrl")
        val host = uri.host?.lowercase()
            ?: throw IllegalArgumentException("baseUrl must contain a host: $baseUrl")
        require(
            scheme == "https" || (scheme == "http" && (isLoopbackHost(host) || allowInsecureHttp))
        ) {
            "baseUrl must use https (http is allowed only for localhost / 127.0.0.1 / 10.0.2.2, " +
                "or any host via the DEV-only allowInsecureHttp flag): $baseUrl"
        }
        require(uri.query == null && uri.fragment == null) {
            "baseUrl must not contain a query or fragment: $baseUrl"
        }
        val port = uri.port
        origin = buildString {
            append(scheme).append("://").append(host)
            if (port != -1 && !isDefaultPort(scheme, port)) append(':').append(port)
        }
        // basePathOf(): path component without trailing slash, '' for root.
        basePath = (uri.path ?: "").trimEnd('/')
        normalizedBaseUrl = baseUrl.trim().trimEnd('/')
        val hosts = HashSet<String>()
        hosts.add(host)
        for (entry in allowedOrigins) {
            val trimmed = entry.trim().lowercase()
            if (trimmed.isEmpty()) continue
            if (trimmed.contains("://")) {
                Uri.parse(trimmed).host?.let { hosts.add(it.lowercase()) }
            } else {
                // Bare host, possibly with a port — strip the port.
                hosts.add(trimmed.substringBefore('/').substringBefore(':'))
            }
        }
        allowedHosts = hosts
        baseScheme = scheme
        baseHost = host
    }

    /**
     * `true` when [url]'s origin is acceptable for main-frame navigation and
     * bridge traffic: an allowed host over https (or http for loopback hosts).
     *
     * Cleartext here rides on the requested [allowInsecureHttp] rather than the
     * build-gated one, which is safe only because [isBaseUrlTransportAllowed]
     * stops a cleartext session before a page exists to navigate. A caller that
     * reaches this without that gate must apply it first.
     */
    fun isUrlOriginAllowed(url: String?): Boolean {
        if (url.isNullOrEmpty()) return false
        val uri = Uri.parse(url)
        val scheme = uri.scheme?.lowercase() ?: return false
        val host = uri.host?.lowercase() ?: return false
        if (scheme != "https" && !(scheme == "http" && (isLoopbackHost(host) || allowInsecureHttp))) {
            return false
        }
        return host in allowedHosts
    }

    /**
     * `true` when the DEV-only [allowInsecureHttp] escape hatch is honoured in
     * [context]'s app.
     *
     * The hatch counts only where the host app itself is debuggable: a release
     * APK must not be talked into cleartext by one boolean in the embedding
     * code. Ignoring it is logged — a portal that silently refuses to load is
     * worse than one that says why.
     */
    fun isInsecureHttpHonoured(context: Context): Boolean {
        if (!allowInsecureHttp) return false
        val debuggable = (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
        if (!debuggable) {
            Log.w(
                TAG,
                "allowInsecureHttp is ignored in a non-debuggable build — " +
                    "the portal must be served over https."
            )
        }
        return debuggable
    }

    /**
     * `true` when [baseUrl] may be loaded in [context]'s app: https, a loopback
     * host, or a debuggable app that opted into [allowInsecureHttp].
     *
     * The constructor accepts a cleartext [baseUrl] whenever [allowInsecureHttp]
     * was passed, because a config is just data and has no Context to judge the
     * build with. THIS is the check that decides whether the request may leave
     * the device, and [AmthalPortalView] makes it before every load.
     */
    fun isBaseUrlTransportAllowed(context: Context): Boolean =
        baseScheme == "https" || isLoopbackHost(baseHost) || isInsecureHttpHonoured(context)

    private companion object {
        private const val TAG = "AmthalPortal"

        private fun isLoopbackHost(host: String): Boolean =
            host == "localhost" || host == "127.0.0.1" || host == "10.0.2.2" || host == "::1"

        private fun isDefaultPort(scheme: String, port: Int): Boolean =
            (scheme == "https" && port == 443) || (scheme == "http" && port == 80)
    }
}
